Source/Base/Base/Math/Spline.cpp

```diff
@@ -0,0 +1,261 @@
+#include "Spline.h"
+#include "Interpolation.h"
+
+namespace Spline
+{
+    InterpolatedStorage::InterpolatedStorage(u32 step)
+        : _stepSize(step)
+    {
+
+    }
+
+    void InterpolatedStorage::SetStep(u32 newStep)
+    {
+        if (newStep == 0)
+            return;
+
+        _stepSize = newStep;
+        _points.clear();
+        _dirty = true;
+    }
+
+    void InterpolatedStorage::StoreInterpolation(const std::vector<vec3>& points, u32 step)
+    {
+        _stepSize = step;
+        _points = points;
+        _dirty = false;
+    }
+
+    void InterpolatedStorage::AddPortion(const std::vector<vec3>& points, i32 position)
+    {
+        if (_stepSize == 0 || points.size() != (_stepSize + 1))
+            return;
+
+        if (position < 0)
+        {
+            _points.insert(_points.end(), points.begin(), points.end());
+            _dirty = false;
+            return;
+        }
+
+        i32 newPosition = position * (_stepSize + 1);
+        if (position < _points.size())
+        {
+            _points.insert(_points.begin() + newPosition, points.begin(), points.end());
+            _dirty = false;
+            return;
+        }
+    }
+
+    const std::vector<vec3>& InterpolatedStorage::GetInterpolatedStorage(u32 splinePoints)
+    {
+        bool checkSize = (splinePoints == 0) ?
+            true : (splinePoints * (_stepSize + 1)) == _points.size();
+
+        if (IsDirty() || !checkSize)
+            return {};
+
+        return _points;
+    }
+
+    Spline::Spline(InterpolationType type)
+        : _type(type),
+        _storage(0)
+    {
+        _storage.Clear();
+        _storage.MarkAsDirty();
+    }
+
+    Spline::Spline(InterpolationType type, const std::vector<SplinePoint>& points)
+        : _type(type),
+        _storage(0)
+    {
+        _points = points;
+
+        _storage.Clear();
+        _storage.MarkAsDirty();
+    }
+
+    void Spline::AddPoint(const SplinePoint& newPoint, i32 position)
+    {
+        if (position < 0)
+        {
+            _points.emplace_back(newPoint);
+        }
+        else
+        {
+            _points.insert(_points.begin() + position, newPoint);
+        }
+
+        _storage.MarkAsDirty();
+    }
+
+    void Spline::RemovePoint(u32 position)
+    {
+        if (position >= _points.size())
+            return;
+
+        std::vector<SplinePoint>::iterator it = _points.begin();
+        std::advance(it, position);
+        _points.erase(it);
+        _storage.MarkAsDirty();
+    }
+
+    void Spline::SetStep(u32 newStep)
+    {
+        newStep = Math::Max((i32)newStep, 1);
+
+        if (_step == newStep)
+            return;
+
+        _step = newStep;
+        _storage.SetStep(newStep);
+    }
+
+    void Spline::SetType(InterpolationType newType)
+    {
+        if (newType < InterpolationType::None || newType >= InterpolationType::MAX_TYPE || _type == newType)
```

Comment:
This is hard to read (Either make variables above that show case what is being checked, or have 2 if statements that each deal with a respective sanity check)